JavaScript/queuing-system.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+
+function Queue(userName) {
+  this.userName = userName;
+  this.first = null;
+  this.last = null;
+  this.length = 0;
+  this.timeout = 0;
+}
+
+Queue.prototype.setTimeout = function(msec) {
+  this.timeout = msec;
+  return this;
+};
+
+Queue.prototype.add = function(item, priority) {
+  let element = { prev: null, next: null, item, priority };
+  if (!this.last) {
+    this.first = element;
+    this.last = element;
+  } else {
+    let current = this.last;
+    for (let i = 1; i < this.length; i++) {
+      if (current.priority <= element.priority) current = current.prev;
+    }
+    if (!current.prev && current.priority < element.priority) {
+      element.next = this.first;
+      this.first.prev = element;
+      this.first = element;
+    } else if (current.priority >= element.priority) {
+      element.next = current.next;
+      current.next = element;
+      element.prev = current;
+      if (!element.next) this.last = element;
+    }
+  }
+  this.length++;
+  if (this.timeout) {
+    setTimeout(() => {
+      if (this.length > 1) {
+        this.last = element.prev;
+        element = null;
```

Comment:
Why?

Reply:
After some time I delete the last element, if it is still in the queue

Reply:
Oh, it's my fault. I guess it should be:
this.last = this.last.prev;
this.last.next = null;